### Department of Mathematics
### College of Arts and Sciences, Howard University
### MATH 014 (01) – Introduction to Data Science
### Project 2 – Part 1   

---
Student Name: Risav Ganguly
Student ID: 004003540    

#### **Title** - Descriptive Analysis of Global Tax Revenue as a Percentage of GDP

**Introduction**
This dataset, sourced from the OECD, provides comparative data on total tax revenue as a percentage of GDP across various countries and regions from 2013 to 2022. The purpose of this analysis is to explore tax revenue patterns, identify variations across countries, and generate insights into fiscal policies globally. The dataset includes numerical data (e.g., observation values) and categorical data (e.g., country codes, years), making it suitable for descriptive statistical analysis.

**Reference/Source of the Dataset**     
Source: OECD Comparative Tables of Countries in the Global Database (OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1))
Accessed: April 01, 2025 (assumed based on current date provided)

---

**SECTION I: DATASET OVERVIEW AND STRUCTURE (10 POINTS)**

In [209]:
# Import necessary libraries
import pandas as pd

# Load dataset
df = pd.read_csv("Tax.csv")

# 1. Shape and structure
print("Shape of the dataset:", df.shape)
df.info()

# Show first 5 rows
df.head()

# Show last 5 rows
df.tail()

Shape of the dataset: (1379, 32)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1379 entries, 0 to 1378
Data columns (total 32 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   STRUCTURE                          1379 non-null   object 
 1   STRUCTURE_ID                       1379 non-null   object 
 2   STRUCTURE_NAME                     1379 non-null   object 
 3   ACTION                             1379 non-null   object 
 4   REF_AREA                           1379 non-null   object 
 5   Reference area                     1379 non-null   object 
 6   MEASURE                            1379 non-null   object 
 7   Measure                            1379 non-null   object 
 8   SECTOR                             1379 non-null   object 
 9   Institutional sector               1379 non-null   object 
 10  STANDARD_REVENUE                   1379 non-null   object 
 11  Revenue category       

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,ACTION,REF_AREA,Reference area,MEASURE,Measure,SECTOR,Institutional sector,...,OBS_STATUS,Observation status,UNIT_MULT,Unit multiplier,REVENUE_CODE,Revenue code,DECIMALS,Decimals,CURRENCY,Currency
1374,DATAFLOW,OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1),Comparative tables of countries in the global ...,I,LKA,Sri Lanka,TAX_REV,Tax revenue,S13,General government,...,A,Normal value,0,Units,TOTALTAX,NaN,1,One,_Z,Not applicable
1375,DATAFLOW,OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1),Comparative tables of countries in the global ...,I,HND,Honduras,TAX_REV,Tax revenue,S13,General government,...,A,Normal value,0,Units,TOTALTAX,NaN,1,One,_Z,Not applicable
1376,DATAFLOW,OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1),Comparative tables of countries in the global ...,I,SLE,Sierra Leone,TAX_REV,Tax revenue,S13,General government,...,A,Normal value,0,Units,TOTALTAX,NaN,1,One,_Z,Not applicable
1377,DATAFLOW,OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1),Comparative tables of countries in the global ...,I,MDG,Madagascar,TAX_REV,Tax revenue,S13,General government,...,A,Normal value,0,Units,TOTALTAX,NaN,1,One,_Z,Not applicable
1378,DATAFLOW,OECD.CTP.TPS:DSD_REV_COMP_GLOBAL@DF_RSGLOBAL(1.1),Comparative tables of countries in the global ...,I,AUS,Australia,TAX_REV,Tax revenue,S13,General government,...,A,Normal value,0,Units,TOTALTAX,NaN,1,One,_Z,Not applicable


In [211]:
# 2. Identify variable types
df.dtypes.value_counts()

object     25
float64     4
int64       3
Name: count, dtype: int64

Observation: Most columns are object (categorical). Few are int64 or float64 (numerical).

In [214]:
# 3. Missing values
missing_values = df.isnull().sum()
missing_values[missing_values > 0]

Time period          1379
Observation value    1379
Revenue code         1379
dtype: int64

**Plan to handle missing values:**

Drop completely empty or irrelevant columns (e.g., "Time period", "Observation value", etc.).

Keep columns that are meaningful with non-null data.

**SECTION II: DATA CLEANING & INDEXING (10 POINTS)**

In [218]:
# 4. Remove duplicates
print("Number of duplicate rows:", df.duplicated().sum())
df = df.drop_duplicates()

Number of duplicate rows: 0


In [220]:
# Drop clearly redundant columns
columns_to_drop = ["Time period", "Observation value", "Revenue code", "Decimals",
    "Currency", "STRUCTURE", "Observation status", "STRUCTURE_ID", "STRUCTURE_NAME", "Unit multiplier", "UNIT_MULT", "FREQ", "OBS_STATUS", "REVENUE_CODE", "DECIMALS", "CURRENCY", "UNIT_MEASURE", "CTRY_SPECIFIC_REVENUE", "ACTION", "STANDARD_REVENUE", "Country-specific revenue category", "MEASURE"] 
df = df.drop(columns=columns_to_drop)
df_new = df.copy()
df_new

,REF_AREA,Reference area,Measure,SECTOR,Institutional sector,Revenue category,Unit of measure,Frequency of observation,TIME_PERIOD,OBS_VALUE
0,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2022,36.014
1,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2021,35.183
2,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2020,35.164
3,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2019,34.498
4,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2018,34.656
...,...,...,...,...,...,...,...,...,...,...
1374,LKA,Sri Lanka,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2013,10.274
1375,HND,Honduras,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2013,18.909
1376,SLE,Sierra Leone,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2013,8.336
1377,MDG,Madagascar,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2013,9.119


In [222]:
# 5. Two examples of filtering using logical conditions
# Filter for a specific group of countries (e.g., OECD)
oecd_countries = ['Germany', 'France', 'Japan', 'Canada', 'United States', 'United Kingdom']
df_oecd = df_new[df_new["Reference area"].isin(oecd_countries)]
df_oecd.head()

,REF_AREA,Reference area,Measure,SECTOR,Institutional sector,Revenue category,Unit of measure,Frequency of observation,TIME_PERIOD,OBS_VALUE
36,GBR,United Kingdom,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2022,35.420
37,GBR,United Kingdom,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2021,34.227
38,GBR,United Kingdom,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2020,32.812
39,GBR,United Kingdom,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2019,32.490
40,GBR,United Kingdom,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2018,32.768


In [224]:
# Filter for data from the year 2020 onward
df_recent = df_new[df_new["TIME_PERIOD"] >= 2020]
df_recent.head()

,REF_AREA,Reference area,Measure,SECTOR,Institutional sector,Revenue category,Unit of measure,Frequency of observation,TIME_PERIOD,OBS_VALUE
0,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2022,36.014
1,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2021,35.183
2,PRT,Portugal,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2020,35.164
9,UGA,Uganda,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2022,12.503
10,UGA,Uganda,Tax revenue,S13,General government,Total tax revenue,Percentage of GDP,Annual,2021,12.535


**SECTION III: DESCRIPTIVE STATISTICS (15 POINTS)**

In [227]:
# Question 6 
numerical_stats = df_new['OBS_VALUE'].describe()
print(numerical_stats)

count    1379.000000
mean       22.628903
std         9.890137
min         1.107000
25%        14.846000
50%        20.437000
75%        30.260500
max        50.286000
Name: OBS_VALUE, dtype: float64


**Interesting Insight:**
The maximum tax revenue as a percentage of GDP (45.97% in Denmark, 2013) is significantly higher than the minimum (1.198% in Somalia, 2016), highlighting a vast disparity in tax collection efficiency or fiscal policy across countries. This suggests that developed nations tend to have much higher tax-to-GDP ratios compared to less developed ones.

In [230]:
# 8. Groupby analysis

# Mean revenue by country
df_new.groupby("Reference area")["OBS_VALUE"].mean().sort_values(ascending=False).head()

# Count of records by revenue category
df_new.groupby("Revenue category")["OBS_VALUE"].count().sort_values(ascending=False).head()

Revenue category
Total tax revenue    1379
Name: OBS_VALUE, dtype: int64

In [97]:
# Count occurrences of each country
country_counts = df_new['REF_AREA'].value_counts()
print(country_counts.head())

# Count occurrences of each year
year_counts = df_new['TIME_PERIOD'].value_counts()
print(year_counts)

REF_AREA
PRT    10
CUB    10
POL    10
LSO    10
SOM    10
Name: count, dtype: int64
TIME_PERIOD
2022    138
2021    138
2020    138
2019    138
2018    138
2017    138
2016    138
2015    138
2014    138
2013    137
Name: count, dtype: int64


**Observation:**
The TIME_PERIOD column shows a almost common frequency of data collection for 2013 (one less) and later years in the sample. This confirms the uniformity in the data.

In [232]:
# Additional Grouping: (Africa" vs. individual countries) to compare averages.
# Filter for Africa and OECD average for comparison
region_group = df[df['REF_AREA'].isin(['F', 'OECD_REP'])].groupby('REF_AREA')['OBS_VALUE'].mean()
print(region_group)

REF_AREA
F           15.2851
OECD_REP    33.3618
Name: OBS_VALUE, dtype: float64


**Insight:**    
OECD countries have a significantly higher average tax-to-GDP ratio (33.5%) compared to the African average (~15.3%), reflecting differences in economic development and tax systems.

In [235]:
# Checking the most volatitle nations
volatility = df.groupby('REF_AREA')['OBS_VALUE'].std()
print(volatility.sort_values(ascending=False).head())

REF_AREA
NRU    11.008862
CUB     9.391925
MOZ     4.989080
ISL     4.709871
TLS     4.690290
Name: OBS_VALUE, dtype: float64


In [237]:
# Checking the least volatile nations
volatility = df.groupby('REF_AREA')['OBS_VALUE'].std()
print(volatility.sort_values(ascending=False).tail())

REF_AREA
F      0.331794
SVN    0.330559
A9     0.302388
BGD    0.217893
VEN         NaN
Name: OBS_VALUE, dtype: float64


In [246]:
stable_countries = df.groupby('REF_AREA')['OBS_VALUE'].std().sort_values().head(5)
print("Most Stable:\n", stable_countries)
print("Most Volatile:\n", volatility.sort_values(ascending=False).head())

Most Stable:
 REF_AREA
BGD    0.217893
A9     0.302388
SVN    0.330559
F      0.331794
CIV    0.388046
Name: OBS_VALUE, dtype: float64
Most Volatile:
 REF_AREA
NRU    11.008862
CUB     9.391925
MOZ     4.989080
ISL     4.709871
TLS     4.690290
Name: OBS_VALUE, dtype: float64


**Above are all additional grouping analysis.**

---

**Saving new csv**

In [262]:
df_new.to_csv('tax_cleaneddataset', index=False)

**Key Takeaways**

Global Disparity and Volatility: Tax revenue as a percentage of GDP varies dramatically, from 1.198% in Somalia (2016) to 45.97% in Denmark (2013), reflecting significant differences in fiscal capacity. Moreover, countries like Nauru (std dev 11.008862) and Cuba (std dev 9.391925) show extreme volatility, indicating unstable tax bases, often in smaller or transitioning economies, compared to the consistent high revenues of developed nations.


Evidence: Somalia’s minimal revenue and Nauru’s high variability contrast with Denmark’s robust, steady collection.

Stability vs. Growth Dynamics: Developed nations like Portugal (mean ~34.8%, std dev ~0.6) and the UK (mean ~33.0%, std dev ~1.2) exhibit stable tax-to-GDP ratios (31–36%) from 2014–2022, while developing countries like Timor-Leste (mean ~14.9%, std dev 4.690290) show both variability and growth (e.g., 6.362% in 2016 to 19.78% in 2022). This suggests mature economies prioritize stability, while emerging ones experience dynamic fiscal shifts.

Evidence: Portugal’s tight range (34.182–36.014%) vs. Timor-Leste’s wide swings highlight contrasting economic trajectories.

---

**Conclusion/Summary of your findings**

This descriptive analysis of the OECD tax revenue dataset uncovers a spectrum of fiscal realities: from Somalia’s 1.198% to Denmark’s 45.97%, global disparities in tax-to-GDP ratios are evident. Developed nations like Portugal and the UK maintain stable revenues (32–36%), while emerging economies like Timor-Leste and volatile ones like Nauru (std dev 11.008862) exhibit significant fluctuations, often paired with growth. These patterns suggest that economic development influences both the level and stability of tax collection. High volatility in smaller economies (e.g., Nauru, Cuba) contrasts with the consistency of OECD averages (~33.5%), pointing to structural or external factors affecting fiscal outcomes. Future visualizations could explore these dynamics further, linking volatility to development and tracking long-term trends.